contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1112 @@
+            msg.sender,
+            order.erc721TokenId
+        );
+
+        if (takerCallbackData.length > 0) {
+            require(
+                msg.sender != address(this), 
+                "ERC721OrdersFeature::_buyERC721/CANNOT_CALLBACK_SELF"
+            );
+            uint256 ethBalanceBeforeCallback = address(this).balance;
+            // Invoke the callback
+            bytes4 callbackResult = IERC721OrderCallback(msg.sender)
+                .zeroExERC721OrderCallback(takerCallbackData);
+            ethAvailable = ethAvailable.safeAdd(
+                address(this).balance.safeSub(ethBalanceBeforeCallback)
+            );
+            // Check for the magic success bytes
+            require(
+                callbackResult == TAKER_CALLBACK_MAGIC_BYTES,
+                "ERC721OrdersFeature::_sellERC721/CALLBACK_FAILED"
+            );            
+        }
+
+        if (address(order.erc20Token) == NATIVE_TOKEN_ADDRESS) {
+            // Check that we have enough ETH.
+            rrequire(
+                ethAvailable >= order.erc20TokenAmount,
+                LibERC721OrdersRichErrors.InsufficientEthError(
+                    ethAvailable,
+                    order.erc20TokenAmount
+                )
+            );
+            // Transfer ETH to the seller.
+            _transferEth(payable(order.maker), order.erc20TokenAmount);
+            // Fees are paid from the EP's current balance of ETH.
+            uint256 ethFees = _payFees(order, address(this), true);
+            // Sum the amount of ETH spent.
+            ethSpent = order.erc20TokenAmount.safeAdd(ethFees);
+        } else if (order.erc20Token == WETH) {
+            // If there is enough ETH available, fill the WETH order 
+            // (including fees) using that ETH. 
+            // Otherwise, transfer WETH from the taker.
+            if (ethAvailable >= order.erc20TokenAmount) {
+                // Wrap ETH.
+                WETH.deposit{value: order.erc20TokenAmount}();
+                // TODO: Probably safe to just use WETH.transfer for some
+                //       small gas savings
+                // Transfer WETH to the seller.
+                _transferERC20Tokens(
+                    WETH,
+                    order.maker,
+                    order.erc20TokenAmount
+                );
+                // Pay fees using ETH. 
+                uint256 ethFees = _payFees(order, address(this), true);
+                // Sum the amount of ETH spent.
+                ethSpent = order.erc20TokenAmount.safeAdd(ethFees);
+            } else {
+                // Transfer WETH from the buyer to the seller.
+                _transferERC20TokensFrom(
+                    order.erc20Token,
+                    msg.sender,
+                    order.maker,
+                    order.erc20TokenAmount
+                );
+                // The buyer pays fees using WETH.
+                _payFees(order, msg.sender, false);
```

Comment:
can this be refactored to fall through to the dupe code below to reduce LoC? :sweat_smile: